### PR TITLE
[WFLY-15685] Jakarta Concurrency 3.0 in WildFly Preview

### PR DESCRIPTION
**Pull request by emmartins**
Issue: https://issues.redhat.com/browse/WFLY-15685
Important Note: There are multiple FIXMEs and TODOs, for instance to extract followups, or simply add new DUP phase values to WFLY Core, but I believe we should deal with those later, once all reviewers agree this is the way to go and everything else is good...

### CODE REVIEW

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/logging/EeLogger.java
Comment (bstansberry) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/logging/EeLogger.java:1219]: s/usage/uses

**Comment by bstansberry** [comment]
As for duplicating the ee subsystem source, I can live with that for now, but we need to follow up with a refactor of ee-concurrency into a separate maven module, so we can limit the duplication to those parts.

**Comment by bstansberry** [comment]
BTW, I've been looking at the 3rd commit (the main one) but haven't finished.

**Comment by emmartins** [comment]
> All the stuff in the transaction subsystem should be moved to the ee subsystem. Old stuff to ee, new to ee-9/source-transform/ee. And then there is no reason for duplicating the tx subsystem code.
> 
> The only reason the org.jboss.as.txn.ee.concurrency package is in the tx subsystem is to use TransactionManagerService.INTERNAL_SERVICE_NAME to inject a TransactionManager. That should not be necessary.
> 
> In the ee subsystem code:
> 
> 1. Check for capability "org.wildfly.transactions.global-default-local-provider". If not present the transaction subsystem is not present, and skip the rest.
> 2. Don't inject the TM. Just add a non-injection requirement for the "org.wildfly.transactions.global-default-local-provider" capability. This ensures the tm stuff is ready before your service.
> 3. Instead of using injection of the TM, directly use org.wildfly.transaction.client.ContextTransactionManager.getInstance(). That's what TransactionManagerService.getValue() provides, so doing it directly is the same as injecting it.
> 
> Note that the above is the way access to the TM is meant to be handled; it's not a hack to avoid forking. See comment above TransactionManagerService.SERVICE_NAME.

Hi @bstansberry , to avoid making this even more complex I created a separated JIRA and PR for this task at https://github.com/wildfly/wildfly/pull/15811

Once that PR is approved and merged I will rebase and update this PR.

PS: I believe you can still continue to review the 3rd commit on this one.

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/deployers/EEConcurrentContextProcessor.java
Comment (jamezp) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/deployers/EEConcurrentContextProcessor.java:118]: This iteration should likely be wrapped in a privileged action if the security manager is present. It looks like it uses the modules class loader which could be a deployment.

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/handle/EE10ContextHandleFactory.java
Comment (jamezp) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/handle/EE10ContextHandleFactory.java:28]: We should come up with another name for this and avoid "EE10".

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ContextServiceDefinitionDescriptorProcessor.java
Comment (jamezp) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ContextServiceDefinitionDescriptorProcessor.java:59]: What does this mean?

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedExecutorDefinitionInjectionSource.java
Comment (jamezp) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedExecutorDefinitionInjectionSource.java:85]: This could be right as I'm not real familiar with how this works in WildFly. However, the `ManagedExecutorDefinition.context()` default seems like it should be used. It does seem to be "java:comp/DefaultContextService", but it just seems a bit fragile to hard-code it.

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedScheduledExecutorDefinitionInjectionSource.java
Comment (jamezp) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedScheduledExecutorDefinitionInjectionSource.java:84]: Same comment here as the previous one.

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedThreadFactoryDefinitionInjectionSource.java
Comment (jamezp) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedThreadFactoryDefinitionInjectionSource.java:61]: Same here too :)

**Review by jamezp — COMMENTED** [review]
This is just a partial review of the "[adds support for the new major features of EE Concurrency 3.0:](https://github.com/wildfly/wildfly/pull/15807/commits/bba111c81b8a06578330e93baeef3e9003d19326)". I will do the others later today.

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ContextServiceDefinitionDescriptorProcessor.java
Comment (emmartins) [reply ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ContextServiceDefinitionDescriptorProcessor.java:59]: XML elements have generic properties, I will need to followup on this wrt propagating such properties to the injectionsource and beyond (if makes any sense)

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedExecutorDefinitionInjectionSource.java
Comment (emmartins) [reply ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedExecutorDefinitionInjectionSource.java:85]: Not sure I am following, you would prefer to grab the JNDI address for the platform's default ContextService from somewhere else?

Comment (jamezp) [reply ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedExecutorDefinitionInjectionSource.java:85]: Sorry. I just mean the `ManagedExecutorDefinition` annotation has defaults and it just feels a little cleaner to use the default rather than the hard-coded version. The hard-coded string does match the default so it's not critical, but if the default were to ever change we could see potential issues.

**Comment by jamezp** [comment]
> As for duplicating the ee subsystem source, I can live with that for now, but we need to follow up with a refactor of ee-concurrency into a separate maven module, so we can limit the duplication to those parts.

My concern here is that we would have to fix everything twice. Once in `ee` and once in `ee-9/source-transform/ee`.

**Comment by bstansberry** [comment]
@emmartins https://issues.redhat.com/browse/WFLY-16657 is merged / resolved. Thanks!

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/logging/EeLogger.java
Comment (emmartins) [reply ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/logging/EeLogger.java]: fixed

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedExecutorDefinitionInjectionSource.java
Comment (emmartins) [reply ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/resource/definition/ManagedExecutorDefinitionInjectionSource.java]: Changed to the const we had in the deployer responsible for binding it.

**Comment by emmartins** [comment]
@bstansberry @jamezp PR reworked on top of [WFLY-16657](https://issues.redhat.com/browse/WFLY-16657), and also addressing your previous concerns. I also left a few "TODO *FOLLOW UP*", identifying the needed followups (on new JIRAs), if this is approved as it is.

File: ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/ContextServiceImpl.java
Comment (bstansberry) [inline ee-9/source-transform/ee/src/main/java/org/jboss/as/ee/concurrent/ContextServiceImpl.java:35]: I'm not asking you to fix this, but the old order is correct per our codestyle:

import static

import java
import javax

import other

**Review by bstansberry — APPROVED** [review]
@emmartins Looks good. Please add the JIRA to the commit messages. If you push and the only changes are to the commit messages, please note that in a PR comment in case we don't want to wait for CI.

**Comment by emmartins** [comment]
@bstansberry JIRA key added to all commits, no other changes done

**Pull request merged**
